src/common/proxy_wasm/ngx_proxy_wasm_foreign_call.c

```diff
@@ -0,0 +1,308 @@
+                                           NGX_PROXY_WASM_ACTION_PAUSE);
+
+    } else {
+        ngx_log_debug0(NGX_LOG_DEBUG_WASM, pwexec->log, 0,
+                       "proxy_wasm last foreign function callback handled");
+
+        ngx_wasm_continue(&call->rctx->env);
+        ngx_proxy_wasm_ctx_set_next_action(pwexec->parent,
+                                           NGX_PROXY_WASM_ACTION_CONTINUE);
+
+        ngx_proxy_wasm_resume(pwexec->parent, pwexec->parent->phase, step);
+    }
+
+    ngx_proxy_wasm_foreign_call_destroy(call);
+}
+
+
+static void
+ngx_proxy_wasm_hfuncs_resolve_lua_handler(ngx_resolver_ctx_t *rslv_ctx)
+{
+#if (NGX_HAVE_INET6)
+    struct sockaddr_in6            *sin6;
+#endif
+    struct sockaddr_in             *sin;
+    u_char                         *p;
+    u_short                         sa_family = AF_INET;
+    ngx_str_t                       args;
+    ngx_buf_t                      *b;
+    ngx_wasm_socket_tcp_t          *sock = rslv_ctx->data;
+    ngx_wasm_lua_ctx_t             *lctx = sock->lctx;
+    ngx_proxy_wasm_dispatch_op_t   *dop = sock->data;
+    ngx_proxy_wasm_foreign_call_t  *call = dop->call.foreign;
+    u_char                          buf[rslv_ctx->name.len +
+#if (NGX_HAVE_INET6)
+                                        sizeof(struct in6_addr) + 1];
+#else
+                                        sizeof(struct in_addr) + 1];
+#endif
+
+    ngx_memzero(buf, sizeof(buf));
+    p = buf;
+
+    if (rslv_ctx->addr.socklen == sizeof(struct sockaddr_in6)) {
+        sa_family = AF_INET6;
+    }
+
+    if (rslv_ctx->state || !rslv_ctx->naddrs) {
+        p++;  /* buffer's 1st byte is address length; 0 if address not found */
+        goto not_found;
+    }
+
+    ngx_wa_assert(rslv_ctx->naddrs == 1);
+
+    switch (sa_family) {
+#if (NGX_HAVE_INET6)
+    case AF_INET6:
+          sin6 = (struct sockaddr_in6 *) rslv_ctx->addr.sockaddr;
+          *(p++) = sizeof(struct in6_addr);
+          p = ngx_cpymem(p, &sin6->sin6_addr, sizeof(struct in6_addr));
+          break;
+#endif
+    default: /* AF_INET */
+          sin = (struct sockaddr_in *) rslv_ctx->addr.sockaddr;
+          *(p++) = sizeof(struct in_addr);
+          p = ngx_cpymem(p, &sin->sin_addr, sizeof(struct in_addr));
+    }
+
+not_found:
+
+    p = ngx_cpymem(p, rslv_ctx->name.data, rslv_ctx->name.len);
+    args.data = buf;
+    args.len = p - buf;
+
+    call->args_out = ngx_wasm_chain_get_free_buf(call->pwexec->pool,
+                                                 &call->rctx->free_bufs,
+                                                 args.len, buf_tag, 1);
+    if (call->args_out == NULL) {
+        goto error;
+    }
+
+    b = call->args_out->buf;
+    b->last = ngx_cpymem(b->last, args.data, args.len);
+
+    if (lctx->yielded) {
+        ngx_proxy_wasm_foreign_callback(dop);
+
+        if (rslv_ctx->state == NGX_WASM_LUA_RESOLVE_ERR) {
+            ngx_wasm_resume(&call->rctx->env);
```

Comment:
It beats me why this line and a few other ones in this file are not covered in codecov but definitively are when testing locally. Any ideas @casimiro?

Reply:
The coverage report for this file has puzzled me as well.

My guess is that the nested `#if` directives were not properly processed by Codecov. 

Coveralls [reports](https://coveralls.io/builds/71982730/source?filename=src%2Fcommon%2Fproxy_wasm%2Fngx_proxy_wasm_foreign_call.c) this file as fully covered.